PP-5640: Add exemption request to authorisation order

oswaldquek opened the pull request:
An exemption request should be added to the authorisation order if:
- Worldpay gateway account has exemption_engine set to true in the
  worldpay_3ds_flex_credentials table
- the gateway account has 3-D Secure enabled

This PR actually covers all the scenarios in PP-5640.

alexbishop1 commented on src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProvider.java:
```diff
@@ -221,9 +222,16 @@ private GatewayOrder buildAuthoriseOrder(CardAuthorisationGatewayRequest request
         boolean is3dsRequired = request.getAuthCardDetails().getWorldpay3dsFlexDdcResult().isPresent() ||
                 request.getGatewayAccount().isRequires3ds();
 
+        boolean exemptionEngine = request.getGatewayAccount().isRequires3ds() ? 
+                Optional.ofNullable(request.getGatewayAccount().getWorldpay3dsFlexCredentials())
+                        .map(Worldpay3dsFlexCredentials::isExemptionEngine)
+                        .orElse(false) 
+                : false;
```
This can be rewritten as:

```java
boolean exemptionEngine = request.getGatewayAccount().isRequires3ds() &&
    Optional.ofNullable(request.getGatewayAccount().getWorldpay3dsFlexCredentials())
            .map(Worldpay3dsFlexCredentials::isExemptionEngine)
            .orElse(false);
```

However, I wonder if we actually need this flag to consider whether or not 3DS is enabled. Presumably it’s just going to be used inside the template and we can nest it inside a check that 3DS is enabled.

oswaldquek replied:
It looks like the following works in the template:
```            
            <#if requires3ds && exemptionEngineEnabled>
                <exemption type="OP" placement="AUTHORISATION"/>
            </#if>
```

alexbishop1 commented on src/main/java/uk/gov/pay/connector/gatewayaccount/model/Worldpay3dsFlexCredentials.java:
```diff
@@ -14,13 +14,25 @@
 
     @JsonIgnore
     private String jwtMacKey;
+    private boolean exemptionEngine;
```
Do we need to update the `equals(…)` and `hashCode()` methods here?

alexbishop1 commented on src/main/java/uk/gov/pay/connector/gatewayaccount/model/Worldpay3dsFlexCredentials.java:
```diff
@@ -14,13 +14,25 @@
     }
 
+    public boolean isExemptionEngine() {
+        return exemptionEngine;
```
Looks like this could return null in some cases. Do we want that or should we default it to false?

oswaldquek replied:
I'm not sure... I think it's the Worldpay3dsFlexCredentials that could be null, not the exemption engine flag which is by default false in the database?

alexbishop1 replied:
The existing constructor will leave it as null if it’s invoked (maybe that constructor isn’t required any more).

oswaldquek replied:
Won't it be `false` by default?

alexbishop1 replied:
Yeah, you’re right, it will. But is the constructor that doesn’t set it even needed now?

alexbishop1 commented on src/test/java/uk/gov/pay/connector/gateway/worldpay/WorldpayPaymentProviderTest.java:
```diff
@@ -333,15 +334,119 @@ void should_include_completed_authentication_in_second_order_when_paRequest_not_
+        assertThat(getNodeListFromExpression(document, "/paymentService/submit/order/exemption").getLength(),
+                is(0));
+    }
+
```
Is it feasible to remove some of the duplication in these tests?

oswaldquek replied:
Done

alexbishop1 reviewed (changes requested):
Looks like it should do the job. Just a few small things.

oswaldquek commented on src/main/java/uk/gov/pay/connector/gateway/worldpay/WorldpayOrderRequestBuilder.java:
```diff
@@ -31,6 +31,7 @@
         private String paResponse3ds;
         private String payerIpAddress;
         private String state;
+        private boolean exemptionEngine;
```
👍

oswaldquek commented:
@alexbishop1 Thanks for the review. I've addressed the comments in a few commit to make viewing the changes easier. Happy to merge the commits if everything is ok.

alexbishop1 commented on src/main/java/uk/gov/pay/connector/gatewayaccount/model/Worldpay3dsFlexCredentialsEntity.java:
```diff
@@ -49,6 +52,10 @@ public Worldpay3dsFlexCredentialsEntity(Long gatewayAccountId, String issuer, St
         this.jwtMacKey = jwtMacKey;
     }
 
+    public boolean isExemptionEngine() {
+        return exemptionEngine;
+    }
```
I think this could do with the `Enabled` treatment too.

oswaldquek replied:
Done

alexbishop1 commented on src/main/resources/templates/worldpay/WorldpayAuthoriseOrderTemplate.xml:
```diff
@@ -53,6 +53,9 @@
             />
             </#if>
             </#if>
+            <#if requires3ds && exemptionEngineEnabled>
+                <exemption type="OP" placement="AUTHORISATION"/>
+            </#if>
```
If this was moved a line up so it’s above the `</#if>` that closes the `<#if requires3ds>`, it could just be `<#if exemptionEngineEnabled>`.

oswaldquek replied:
Done

alexbishop1 reviewed (changes requested):
Looks good, just a couple of tiny things.

alexbishop1 reviewed (dismissed):
This looks good. Can you check if the three-arg variant of the `Worldpay3dsFlexCredentials` constructor can be removed? If so, please do that, otherwise I think this is good to go.

oswaldquek commented:
@alexbishop1 Strictly speaking the three-arg constructor can be removed. I didn't because it's used in a few places (mainly tests) so I thought it'd be convenient to leave it.
I've discovered that https://github.com/alphagov/pay-connector/blob/master/src/main/java/uk/gov/pay/connector/gatewayaccount/model/Worldpay3dsFlexCredentialsEntity.java#L44-L50 can be removed though!

alexbishop1 reviewed (approved):
Looks good!
